### PR TITLE
Extract a service component from Add-using independent of the CodeFixProvider

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Features/CSharp/Portable/AddImport/CSharpAddImportCodeFixProvider.cs [hunk 1]
```diff
@@ -147,7 +148,7 @@ internal static class AddImportDiagnosticIds
     }
 
     [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.AddImport), Shared]
-    internal class CSharpAddImportCodeFixProvider : AbstractAddImportCodeFixProvider<SimpleNameSyntax>
+    internal class CSharpAddImportCodeFixProvider : AbstractAddImportCodeFixProvider
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/AddImport/CSharpAddImportCodeFixProvider.cs]: Note: i intentionally spit the types and kept them in the same file to make the PR easier to read.  A following PR will move them into their own files.

File: src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs [hunk 2]
```diff
@@ -41,6 +39,108 @@ internal abstract partial class AbstractAddImportCodeFixProvider<TSimpleNameSynt
...
+
+                context.RegisterFixes(codeActions, diagnostic);
+                codeActions.Free();
+            }
+        }
+
+        private IPackageInstallerService GetPackageInstallerService(Document document)
+            => _packageInstallerService ?? document.Project.Solution.Workspace.Services.GetService<IPackageInstallerService>();
+
+        private CodeAction TryCreateCodeAction(Document document, AddImportFixData fixData)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs]: This is a move.  Just github has no way to show that.

File: src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs [hunk 3]
```diff
@@ -41,6 +39,108 @@ internal abstract partial class AbstractAddImportCodeFixProvider<TSimpleNameSynt
...
+    }
+
+    internal interface IAddImportFeatureService : ILanguageService
+    {
+        Task<ImmutableArray<AddImportFixData>> GetFixesAsync(
+            Document document, TextSpan span, string diagnosticId, ISymbolSearchService symbolSearchService,
+            bool searchReferenceAssemblies, ImmutableArray<PackageSource> packageSources, CancellationToken cancellationToken);
+    }
+
+    internal abstract partial class AbstractAddImportFeatureService<TSimpleNameSyntax>
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs]: Type kept in the same file to make PR easier to review

File: src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs [hunk 4]
```diff
@@ -41,6 +39,108 @@ internal abstract partial class AbstractAddImportCodeFixProvider<TSimpleNameSynt
             _symbolSearchService = symbolSearchService;
         }
 
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs]: this was a general move of this function.  Though some of hte logic around options/services got rejiggered.

**Comment by CyrusNajmabadi** [comment]
CSharpWinForms.ChangeControlProperty failure

**Comment by CyrusNajmabadi** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by CyrusNajmabadi** [comment]
Telling @mattgertz about this.  This separates out the search/analysis portion of Add-Using, and the presentation/user-interaction portion.  The former will move OOP and hte latter will stay in VS.

File: src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs [hunk 5]
```diff
@@ -41,6 +39,111 @@ internal abstract partial class AbstractAddImportCodeFixProvider<TSimpleNameSynt
...
+            var addImportService = document.GetLanguageService<IAddImportFeatureService>();
+
+            var solution = document.Project.Solution;
+            var options = solution.Options;
+
+            var searchReferenceAssemblies = options.GetOption(SymbolSearchOptions.SuggestForTypesInReferenceAssemblies, document.Project.Language);
+            var searchNuGetPackages = options.GetOption(SymbolSearchOptions.SuggestForTypesInNuGetPackages, document.Project.Language);
+
+            var symbolSearchService = searchReferenceAssemblies || searchNuGetPackages
+                ? _symbolSearchService ?? solution.Workspace.Services.GetService<ISymbolSearchService>()
```
Comment (Pilchie) [inline src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs]: Does anything ever set this?

**Pull request merged**